### PR TITLE
Added permanent key mappings

### PR DESCRIPTION
**Pull request by donniebreve**
Added functionality for permanent key remapping. See #15.

### CODE REVIEW

**Review by Adda0 — APPROVED** [review]
LGTM. The remap functionality works flawlessly. And I must say that permanent remapping is definitely a useful option to have.

I would like to see a better explanation of what `remap` does (permanent remapping without the need to hold a hyper key) in the configuration file (as it is with the other table headers (`[...]`)).

I think that for clarity, it is important to have the `remap` table before the hyper key and hyper key bindings tables in the config file to emphasize the independency of permanent mappings to the hyper key.

Furthermore, I think that it is worth to mention in the config file that if you want to define a specific behaviour to a permanently remapped key (`KEY_T`) with a hyper key, you will get a behaviour of a remapped key (`KEY_M`) when pressing the original key (`T`): When pressing hyper key + `t`, you get `e`. I am not sure if this is desirable or whether it would be better to allow explicitly defining a different hyper key behaviour for `KEY_T`: To get `r` when pressing hyper key + `t`.
```toml
[Bindings]
KEY_T=KEY_R
KEY_M=KEY_E

# Permanent remappings
[Remap]
KEY_T=KEY_M
```

As of now, I would approve the PR to be merged. If additional functionality is planned to be added to the PR, I will reevaluate the review.

**Comment by auouymous** [comment]
"Permanent" remapping sounds like it modifies something so the key is forever remapped even if touchcursor is removed.

> How do other key mapping projects (xkb, xmodmap) handle remapped keys. Do later configuration lines still detect the key as 't'?

Xmodmap only binds symbols to keycodes. For `keycode 28 = t T` and `keycode 58 = m M`, the physical `t` key always produces keycode 28 even if you change it to `keycode 28 = m M`. X11 programs that don't read the symbols will bypass any mapping done by Xmodmap.

The proposed example seems the most logical, it produces exactly what you see in the file.

**Comment by donniebreve** [comment]
> Xmodmap only binds symbols to keycodes.

After some cursory research, I noticed this too. I will move forward with the proposed changes, i.e. moving the remap layer to the last layer.

> "Permanent" remapping sounds like it modifies something so the key is forever remapped even if touchcursor is removed.

Any suggestions for a different wording?

**Comment by auouymous** [comment]
> Any suggestions for different wording?

Drop the word "permanent" and only call it key or keycode remapping.

**Comment by Adda0** [comment]
> > Any suggestions for different wording?
> 
> Drop the word "permanent" and only call it key or keycode remapping.

I would rather keep the `Permanent remappings` as the official name for the feature (and therefore keep this in the config file). What I was thinking about is more in the terms of adding additional explanation with examples. Something as follows:
```toml
# Permanent remappings.
# Allows to set permanent remappings when running the service without a need to hold a hyper key.
# Each remapping will be active permanently when the service is running.
# For example, you can swap the functionality of two keys as follows:
# [Remap]
# KEY_T=KEY_M
# KEY_M=KEY_T 
[Remap]
```

**Comment by auouymous** [comment]
Those three comments read exactly the same if "permanent" is removed.

```
# Allows to set remappings when running the service without a need to hold a hyper key.
# Each remapping will be active when the service is running.
# The remapped key can be remapped differently while holding the hyper key in table `[Bindings]` with the original key name:
```

**Comment by Adda0** [comment]
> Those three comments read exactly the same if "permanent" is removed.

Maybe, but if you want to refer to the feature anywhere else without all these exhaustive comments, you have to have a single self-explanatory name for the feature. Without the `permanent` adjective, we have two different features present in the config file: bindings and bindings. That does not help us determine which is which, does it? One is for the hyper key bindings and the second is for permanent bindings. I think using `permanent` makes sense and describes pretty precisely what they are. And no one will think that when the service is not running, the new “permanent” bindings will be active, I believe. And if they do, they have the comments to explain it. 

Furthermore, due to backward compatibility, we cannot rename the hyper key bindings under the table `[Bindings]` unless we want to break everyone's configs. Therefore, we have to add a specific, distinguishable name each time we introduce a new feature (new types of bindings).

File: src/mapper.c [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    }
+    return code;
+}
+
+/**
+ * * Processes a key input event. Converts and emits events as necessary.
+ * */
+void processKey(int type, int code, int value)
+{
+    printf("processKey(in): code=%i value=%i state=%i\n", code, value, state);
```
Comment (Adda0) [inline src/mapper.c:66]: I suppose the `printf` calls in this function should not be in the final production code, right? However, I would keep them here commented out for when I manage to implement debugging options (as in #42). Then, they can be hidden behind `debug active` constraint.

**Review by Adda0 — COMMENTED** [review]
Looks great to me now. The mapping behaves exactly how I would have intuitively understood it to, and every discussed combination of permanent remappings and hyper key remappings works. Even permanent remapping of a key and using the same key as a hyper key.

I would consider this PR to be in a ready-to-merge stage. Up to you to decide whether to undraft the PR and move towards merging. Great job.

**Comment by Adda0** [comment]
@donniebreve What do you think about this PR? I have been using the permanent remapping since opening this PR and I think it works great. I cannot say that I found any problems with it.

As you mentioned somewhere, it might be a good idea to review the whole of `mapper.c` to probably fix and improve the emitting logic, but that would be a task for another PR.

As far as this PR goes, I think it is a great addition to the application and is ready to be merged, except for maybe addressing the debug prints.

File: src/mapper.c (hunk 1 above)
Comment (donniebreve) [reply src/mapper.c]: I commented them out again for now.

**Comment by donniebreve** [comment]
Sure, let's merge this in until I get some motivation to rewrite some things.

**Review by Adda0 — APPROVED** [review]
LGTM. It is a great feature to have (I cannot express enough how much this PR helped me – having enter key in a place of semicolon, so thanks). As agreed upon, I agree that we can merge this PR. Additional changes may come later. Feel free to merge the PR at your leisure.

**Pull request merged**